contracts/extensions/ReputationBootstrapper.sol

```diff
@@ -0,0 +1,194 @@
+  function uninstall() public override auth {
+    uint256 balance = ERC20(token).balanceOf(address(this));
+    require(ERC20(token).transfer(address(colony), balance), "reputation-bootstrapper-transfer-failed");
+
+    selfdestruct(address(uint160(address(colony))));
+  }
+
+  // Public
+
+  /// @notice Set an arbitrary number of grants
+  /// @param _paid An array of booleans indicated pair or unpaid
+  /// @param _hashedSecrets An array of (hashed) secrets
+  /// @param _amounts An array of reputation amounts claimable by the secret
+  function setGrants(bool[] memory _paid, bytes32[] memory _hashedSecrets, uint256[] memory _amounts) public onlyRoot notDeprecated {
+    require(_paid.length == _hashedSecrets.length, "reputation-bootstrapper-invalid-arguments");
+    require(_hashedSecrets.length == _amounts.length, "reputation-bootstrapper-invalid-arguments");
+    uint256 balance = ERC20(token).balanceOf(address(this));
+
+    for (uint256 i; i < _hashedSecrets.length; i++) {
+      require(_amounts[i] <= INT128_MAX, "reputation-bootstrapper-invalid-amount");
+
+      if (_paid[i]) {
+        uint256 priorGrant = grants[_paid[i]][_hashedSecrets[i]].amount;
+        if (priorGrant < _amounts[i]) {
+          totalPaidGrants += _amounts[i] - priorGrant;
+        } else {
+          totalPaidGrants -= priorGrant - _amounts[i];
+        }
+        require(totalPaidGrants <= balance, "reputation-bootstrapper-insufficient-balance");
+      }
+
+      grants[_paid[i]][_hashedSecrets[i]] = Grant(_amounts[i], block.timestamp);
+
+      emit GrantSet(_paid[i], _hashedSecrets[i], _amounts[i]);
+    }
+  }
+
+  /// @notice Commit the secret, beginning the security delay window
+  /// @param _committedSecret A sha256 hash of (userAddress, secret)
+  function commitSecret(bytes32 _committedSecret) public notDeprecated {
+    bytes32 addressHash = keccak256(abi.encodePacked(msgSender(), _committedSecret));
+    committedSecrets[addressHash][_committedSecret] = block.timestamp;
+  }
+
+  /// @notice Claim the grant, after committing the secret and having the security delay elapse
+  /// @param _secret The secret corresponding to a reputation grant
+  function claimGrant(bool _paid, uint256 _secret) public notDeprecated {
+    bytes32 committedSecret = keccak256(abi.encodePacked(msgSender(), _secret));
+    bytes32 addressHash = keccak256(abi.encodePacked(msgSender(), committedSecret));
+    uint256 commitTimestamp = committedSecrets[addressHash][committedSecret];
+
+    require(
+      commitTimestamp > 0 && commitTimestamp + SECURITY_DELAY <= block.timestamp,
+      "reputation-bootstrapper-commit-window-unelapsed"
+    );
+
+    bytes32 hashedSecret = keccak256(abi.encodePacked(_secret));
+    uint256 grantAmount = grants[_paid][hashedSecret].amount;
+    uint256 tokenAmount = _paid ? grants[_paid][hashedSecret].amount : 0;
+    uint256 grantTimestamp = grants[_paid][hashedSecret].timestamp + SECURITY_DELAY; // Don't decay during delay window
```

Comment:
I mean... so what? If everyone is penalized, no-one is penalized. If anything, I would say this gives an advantage to people who are bootstrapped though the extension over those who are awarded reputation directly, as their reputation has decayed for an hour less.

Reply:
Not following your logic -- are people awarded reputation directly decayed by two hours?

I don't feel strongly, if anything I would've expected you to come down on the side of precision, so I can leave this out.

Reply:
If one person is awarded 1000 reputation directly, and another person is awarded 1000 reputation via bootstrapping at the same time (and goes through the claim process immediately), a day later the first person has had their reputation decay for 24 hours, and the second person has had their reputation decay for 23 hours, and so has more influence than the first, which (seems to me) was quite clearly not the intention of the admin.

I think maybe we're just not agreeing on when the 'clock' should start, but in my opinion it's pretty clear that it should be when it's put anywhere in to the system. Otherwise, it becomes very hard for a user to calibrate the reputation they're awarding correctly if they're using the two different processes.